File: Leetcode/Typescript/2570. Merge Two 2D Arrays by Summing Values/solution.ts
// https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/description

function mergeArrays(nums1: number[][], nums2: number[][]): number[][] {
  const result: number[][] = [];
  const idMap = new Map<number, number>();

  for (const [id, num] of nums1) {
    idMap.set(id, (idMap.get(id) || 0) + num);
  }

  for (const [id, num] of nums2) {
    idMap.set(id, (idMap.get(id) || 0) + num);
  }

  const keys = Array.from(idMap.keys());
  keys.sort((a, b) => a - b);

  for (const key of keys) {
    result.push([key, idMap.get(key)!]);
  }

  return result;
}

File: Leetcode/Python/2570. Merge Two 2D Arrays by Summing Values/solution.py
# https://leetcode.com/problems/merge-two-2d-arrays-by-summing-values/description

class Solution:  # Hashmap
    def mergeArrays(self, nums1: List[List[int]], nums2: List[List[int]]) -> List[List[int]]:
        result: List[List[int]] = []

        id_dict = defaultdict(int)

        for id, num in nums1:
            id_dict[id] += num

        for id, num in nums2:
            id_dict[id] += num

        for key in sorted(id_dict.keys()):
            result.append((key, id_dict[key]))

        return result


class Solution:  # Two pointer
    def mergeArrays(self, nums1: List[List[int]], nums2: List[List[int]]) -> List[List[int]]:
        result: List[List[int]] = []

        id_dict = defaultdict(int)

        for id, num in nums1:
            id_dict[id] += num

        for id, num in nums2:
            id_dict[id] += num

        for key in sorted(id_dict.keys()):
            result.append((key, id_dict[key]))

        return result
